### ISSUE TITLE
Function(s) for extracting parts from URLs

### ISSUE THREAD

**Issue by jasonkres** [datalust/seq-tickets#1386]
# Issue

It is difficult to work with URLs. It would help if there was support for using capturing groups in regular expressions (#1383), but even then it would still be very complicated.

# Idea

Provide functions to work with URLs. Such as:

```sql
UrlPart('https://www.example.com:1234/hello/world?alpha=1&beta=2#test', part)
```
* When part is 'scheme' => 'https'
* When part is 'authority' => 'www.example.com:1234'
* When part is 'host' => 'www.example.com'
* When part is 'port' => 1234
* When part is 'path' => '/hello/world'
* When part is 'query' => 'alpha=1&beta=2'
* When part is 'fragment' => 'test'

See: https://en.wikipedia.org/wiki/URL for this terminology.

There are many edge cases in parsing URLs. Suggest cross-testing against a commonly used library.

**Comment by schallm** [comment]
Would love to see something like this implemented, but more general so any regular expression with capturing groups could be used.  

Logs are stored in Postgres database with our current homegrown logging system.  Regular expressions with capture groups can be used when selecting information to display.  ([documentation](https://www.postgresql.org/docs/current/functions-matching.html#:~:text=The%20substring%20function%20with%20two%20parameters%2C%20substring(string%20from%20pattern)))

I would like to use something like this in Seq to show more concise graphs.

Loggers are formatted like namespaces (i.e., "a.b.c.d.f") and I only want "d.f" on the charts so they are readable.

```sql
select 
   -- substring(Logger from '(\w+\.\w+)$') as DisplayLogger, -- Postgres syntax
   count(*) as EventCount
from stream 
where @Timestamp > Now() - 2h
group by Logger
order by EventCount desc
limit 25
```

**Comment by liammclennan** [comment]
@schallm for your specific example I think you might be able to use `indexOf` and `substring`

![image](https://github.com/datalust/seq-tickets/assets/271514/a7e69e4d-7ac6-419c-a495-899671dd9c25)

**Comment by schallm** [comment]
Can’t test right now, but interesting you can use a regex in the `indexof` function. I’ll look into that in the morning. I tried testing with `substring` in my query earlier and I get an error that `substring` can not be used as an aggregate function in a group by query.

**Comment by liammclennan** [comment]
That's right, and there the semantics are the same as postgres. If you group the data then the select clause has to be an aggregate function that can reduce a set of rows (the group) to a single row. This query makes no sense:

![image](https://github.com/datalust/seq-tickets/assets/271514/9b496b78-5ac1-40e0-a333-fd4076ee8800)

You can move the `Substring` operation to the `group by` clause and it will appear in the output. Then choose your aggregate function. 

![image](https://github.com/datalust/seq-tickets/assets/271514/95ee3154-2e92-4714-a0c1-72569083b97b)

All of the [string matching functions](https://docs.datalust.co/docs/scalar-functions#indexoftext-substring) support regular expressions, and the [`ci` modifier](https://docs.datalust.co/docs/the-seq-query-language#string-operations-and-the-ci-modifier).

**Comment by schallm** [comment]
Thanks so much for your help!  This works great!  However, I have a few questions/thoughts. 😄 

> semantics are the same as postgres

While you can group by a substring in Postgres, you have to select it for the columns to be in the results then group by the alias.

Postgres:
```sql
select 
   substring(Logger from '(\w+\.\w+)$') as DisplayLogger,
   count(*) as EventCount
from Events
group by DisplayLogger
order by EventCount desc
```

Working Seq query:
```sql
select count(*) as EventCount
from stream 
where @Timestamp > Now() - 2h
group by substring(Logger, indexof(Logger, /(\w+\.)?\w+$/)) as DisplayLogger
order by EventCount desc
```

For others (and future me), note that you can rename the column in the `group by` statement as shown above.

This works great as substring doesn't need to know the length of the match as I want the remainder of the string.

It would still be nice to have `substring` work directly with regular expressions and a capture group so I could get text from the middle of a string.  For example, I want to get just the milliseconds out of the following string: 

` The process (processId: 23) took 6.41ms to complete. `

Postgres:
```sql
substring(Message from 'took ([\d\.]+)ms')  -- returns '6.41'
```
Could something like the following syntax be added?  Or does functionality similar to this already exist?
```sql
substring(Message, /took ([\d\.]+)ms/) -- returns '6.41'
```

The answer could be "use structured logging".  However, I think being able to extract data out of strings would always be useful.

**Comment by schallm** [comment]
I know this ticket has been open for a while, but would still like regular expression support for substring.

**Comment by liammclennan** [comment]
Hi @schallm and @jasonkres 

You can now do this with the `Replace` function. E.g.

```sql
select Replace('The process (processId: 23) took 6.41ms to complete', /.*took ([\d\.]+)ms.*/, '$1') 
-- 6.41
```

https://datalust.co/blog/replace-function
https://datalust.co/docs/scalar-functions#replacetext-substring-replacement

**Comment by schallm** [comment]
This actually works pretty well.  You do have to match the full string so the replacement works.  Would you be open to a `substring` function with the regex capture?

**Comment by liammclennan** [comment]
@schallm 

What scenario would that address?

**Comment by schallm** [comment]
You are correct that this solution does add the feature requested and it is much appreciated!

However I feel that a new function or `substring` overload would be more discoverable for developers.  I'm not sure I would ever have found this feature if you hadn't pointed me to it.

A new function would not make developers add extra characters in the Regex pattern.  With replace, the pattern must match the full string (most likely `.*` on both sides of the match). 

A new function would not need the replacement parameter at all as it would just extract the captured text.

**Comment by liammclennan** [comment]
Thanks @schallm. I thought that was what you had in mind. Thanks for clarifying.
